gradle-jdks-setup/src/main/java/org/gradle/wrapper/GradleWrapperMain.java

```diff
@@ -0,0 +1,153 @@
+        String[] extraArgs = {
+            // Disable auto-download and auto-detect of JDKs
+            "-Porg.gradle.java.installations.auto-download=false",
+            "-Porg.gradle.java.installations.auto-detect=false",
+            // Set the custom toolchains locations
+            "-Porg.gradle.java.installations.paths=" + String.join(",", toolchains)
+        };
+        String[] combinedArgs =
+                Stream.concat(Arrays.stream(args), Arrays.stream(extraArgs)).toArray(String[]::new);
+        method.invoke(null, new Object[] {combinedArgs});
+    }
+
+    private static List<String> getAndConfigureJdkAutomanagement() throws IOException {
+        // Delegate to gradle-jdks-setup.sh to install the JDK if it hasn't been installed yet
+        Path projectHome = projectHome();
+        if (!isRunningFromGradlew()) {
+            CommandRunner.run(List.of("./gradle/gradle-jdks-setup.sh"), Optional.of(projectHome.toFile()));
+        }
+
+        // Set the daemon Java Home
+        String osName = CurrentOs.get().uiName();
+        String archName = CurrentArch.get().uiName();
+        Path daemonKdkMajorVersionPath = projectHome.resolve("gradle/gradle-daemon-jdk-version");
+        String majorVersion = Files.readString(daemonKdkMajorVersionPath).trim();
+        Path localPathFile = projectHome
+                .resolve("gradle/jdks")
+                .resolve(majorVersion)
+                .resolve(osName)
+                .resolve(archName)
+                .resolve("local-path");
+        String localJdkFileName = Files.readString(localPathFile).trim();
+        Path gradleJdksInstallationDir = getGradleJdksPath();
+        Path jdkInstallationPath = gradleJdksInstallationDir.resolve(localJdkFileName);
+        System.out.println("Setting daemon Java Home to " + jdkInstallationPath.toAbsolutePath());
+        System.setProperty(
+                "org.gradle.java.home", jdkInstallationPath.toAbsolutePath().toString());
+
+        String localPathPattern = String.format("%s/%s/local-path", osName, archName);
+        List<String> toolchainsInstallationPaths = getAllInstallationsPaths(
+                projectHome.resolve("gradle/jdks").toFile(), gradleJdksInstallationDir, localPathPattern);
+        System.out.println("Setting custom toolchains locations to " + toolchainsInstallationPaths);
+        return toolchainsInstallationPaths;
+    }
+
+    private static List<String> getAllInstallationsPaths(
+            File gradleJdkConfigurationDir, Path gradleJdksInstallationDir, String localPathPattern)
+            throws IOException {
+        List<String> paths = new ArrayList<>();
+        File[] files = gradleJdkConfigurationDir.listFiles();
+        if (files == null) {
+            return paths;
+        }
+        for (File file : files) {
+            if (file.isDirectory()) {
+                paths.addAll(getAllInstallationsPaths(file, gradleJdksInstallationDir, localPathPattern));
```

Comment:
It might be better to use `Files.walk` or `Files.visitFileTree` here rather than manually doing the recursive file exploration.

Reply:
In fact, why is this searching the file tree - shouldn't the installation path be at a fixed known location in the `~/.gradle/gradle-jdks/` directory and not some sub-directory?